Function log2(double) in HElib/src/NumbTh.h:321 and /usr/include/x86_64-linux-gnu/bits/mathcalls.h:144 fight over noexcept (or old throw ()) definition.

**Brief Description:**

Having a test.cpp and test.h file that just includes FHE.h results in following error. One can simply make log2(double) in NumbTh.h to be "noexcept", but probably using namespaces and separating the two functions would be better choice. Will propose a quickfix via PullRequest.

Here the output:


~/master_repos/Private-Decision-Tree-Evaluation$ g++ -std=c++17 -O2 -pedantic -Wall -c src/common/test.cpp -I HElib/src -o objects/common/test.o
In file included from HElib/src/DoubleCRT.h:19:0,
                 from HElib/src/FHE.h:19,
                 from src/common/test.h:11,
                 from src/common/test.cpp:8:
HElib/src/NumbTh.h: In function ‘double log2(double)’:
HElib/src/NumbTh.h:321:34: error: declaration of ‘double log2(double)’ has a different exception specifier
 inline double log2(const double x){ return log(x) * 1.442695040889; }
                                  ^
In file included from /usr/include/features.h:367:0,
                 from /usr/include/x86_64-linux-gnu/c++/5/bits/os_defines.h:39,
                 from /usr/include/x86_64-linux-gnu/c++/5/bits/c++config.h:482,
                 from /usr/include/c++/5/new:39,
                 from /usr/local/include/NTL/new.h:5,
                 from /usr/local/include/NTL/tools.h:11,
                 from /usr/local/include/NTL/vector.h:5,
                 from HElib/src/zzX.h:18,
                 from HElib/src/DoubleCRT.h:18,
                 from HElib/src/FHE.h:19,
                 from src/common/test.h:11,
                 from src/common/test.cpp:8:
/usr/include/x86_64-linux-gnu/bits/mathcalls.h:144:1: error: from previous declaration ‘double log2(double) throw ()’
 __MATHCALL (log2,, (_Mdouble_ __x));
 ^
In file included from HElib/src/FHEContext.h:18:0,
                 from HElib/src/FHE.h:20,
                 from src/common/test.h:11,
                 from src/common/test.cpp:8:
HElib/src/PAlgebra.h: In member function ‘long int PAlgebra::ZmStarGen(long int) const’:
HElib/src/PAlgebra.h:141:15: warning: comparison between signed and unsigned integer expressions [-Wsign-compare]
   {  return (i<gens.size())? gens[i] : 0; }
               ^

--- a/src/NumbTh.h
+++ b/src/NumbTh.h
@@ -318,7 +318,7 @@ void applyLinPoly(NTL::GF2E& beta, const NTL::vec_GF2E& C, const NTL::GF2E& alph
 
 //! Base-2 logarithm
 inline double log2(const NTL::xdouble& x){ return log(x) * 1.442695040889; }
-inline double log2(const double x){ return log(x) * 1.442695040889; }
+inline double log2(const double x) noexcept{ return log(x) * 1.442695040889; }
 
 //! @brief Factoring by trial division, only works for N<2^{60}, only the
 //! primes are recorded, not their multiplicity.

